idx/memory/meta_tags.go

```diff
@@ -51,10 +51,10 @@ var (
 type metaTagIdx struct {
 	sync.RWMutex
 	byOrg    map[uint32]*orgMetaTagIdx
-	idLookup func(uint32, tagquery.Query, chan schema.MKey)
+	idLookup func(uint32, tagquery.Query, func([]schema.MKey))
```

Comment:
Could `idLookup` just return the `[]schema.MKey` directly rather than via a callback? The fundamental difference is whether the callback will be executed under the same memory index read lock held during the id lookup or not.

It seems everywhere `idLookup` is used, the result is processed asynchronously anyway, not under the read lock.

Reply:
the main reason why i didn't make it return `[]schema.MKey` is indeed because the result needs to be submitted into the enricher's event queue before releasing the main memory index lock. While working on this I realized that if we first release the main memory index lock, and then push the query result into the event queue, then there are possible scenarios which would lead to a corrupted state in the enricher, for example:

1) a meta record gets added
2) we lookup all ids matching this meta record and keep the result in some variable, now we release the index lock
3) a metric which matches the new meta record gets added to the index and also gets added to enricher. in the enricher it gets enriched without taking the new meta record from 2) into account because it's not in the enricher yet
4) now the new meta record with its metric ids which have been looked up at 2) get added to the enricher

the result is that the metric which got added at 3) will never be enriched, even though the meta record which got added at 2) should match it.

it's true that all the enricher events get processed asynchronously, so the enrichment might happen with a slight delay, but we need to guarantee that the eventual state (when the enricher queue is fully processed) is correct. for this we need to ensure that the main index can't get modified between the id lookup stage of adding/remove meta records and pushing the results into the enricher event queue, because we rely on the processing order to be correct.